Source/Core/Core/HW/SI/SI_DeviceAMBaseboard.cpp

```diff
@@ -0,0 +1,2358 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/HW/SI/SI_DeviceAMBaseboard.h"
+
+#include <algorithm>
+#include <string>
+
+#include <fmt/format.h>
+
+#include "Common/Buffer.h"
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "Common/IOFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/MsgHandler.h"
+#include "Common/Swap.h"
+
+#include "Core/Boot/Boot.h"
+#include "Core/BootManager.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/DVD/AMMediaboard.h"
+#include "Core/HW/DVD/DVDInterface.h"
+#include "Core/HW/EXI/EXI.h"
+#include "Core/HW/GCPad.h"
+#include "Core/HW/MMIO.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/ProcessorInterface.h"
+#include "Core/HW/SI/SI.h"
+#include "Core/HW/SI/SI_Device.h"
+#include "Core/HW/SI/SI_DeviceGCController.h"
+#include "Core/HW/SystemTimers.h"
+#include "Core/Movie.h"
+#include "Core/NetPlayProto.h"
+#include "Core/System.h"
+
+#include "InputCommon/GCPadStatus.h"
+
+namespace SerialInterface
+{
+
+void JVSIOMessage::Start(int node)
+{
+  m_last_start = m_ptr;
+  const u8 hdr[3] = {0xE0, (u8)node, 0};
+  m_csum = 0;
+  AddData(hdr, 3, 1);
+}
+
+void JVSIOMessage::AddData(const u8* dst, size_t len, int sync = 0)
+{
+  if (m_ptr + len >= 0x80)
+  {
+    PanicAlertFmt("JVSIOMessage overrun!");
+    return;
+  }
+
+  while (len--)
+  {
+    const u8 c = *dst++;
+    if (!sync && ((c == 0xE0) || (c == 0xD0)))
+    {
+      m_msg[m_ptr++] = 0xD0;
+      m_msg[m_ptr++] = c - 1;
```

Comment:
Because `m_ptr` can be incremented twice here, the length check is insufficient.